### ISSUE TITLE
 App.on_pause / on_resume on Android 5.1 not called (ref #3202 for iOS)

### ISSUE THREAD

**Issue by marystern** [kivy/kivy#3584]
I have the same problem as #3202 but on Android 5.1 (rather than iOS) ie on_pause() does not get called when it should.

I've just tried this again with current master (commit 0c7cd9ea95851fbe3bbf029a9832232a774fbd5a) and again with the patch from mixx (in #3202) applied: in both cases this is still broken on Android.

My app is a simple one that uses the camera: it correctly calls out to the Android Camera app, but will not return due to this bug.

Here's a snippet of the logs, and I have logging in my on_pause/on_resume() methods which never get called (on_pause does return True by the way):

```
I/python  (13908): [INFO              ] [XXXXXXXX    ] === App STARTED =======================
I/python  (13908): [INFO              ] [OSC         ] using <thread> for socket
I/python  (13908): [INFO              ] [Base        ] Start application main loop
I/python  (13908): [INFO              ] [Android     ] found 17 joystick
I/python  (13908): [INFO              ] [Android     ] create joystick <0>
I/python  (13908): [INFO              ] [Android     ] discard joystick <0> cause no button
...
I/python  (13908): [INFO              ] [Android     ] create joystick <16>
I/ActivityManager(  472): START u0 {act=android.media.action.IMAGE_CAPTURE flg=0x3 cmp=com.google.android.GoogleCamera/com.android.camera.CaptureActivity (has clip) (has extras)} from uid 10155 on display 0
V/WindowManager(  472): addAppToken: AppWindowToken{3d8ce58c token=Token{25f1babf ActivityRecord{214aa8de u0 com.google.android.GoogleCamera/com.android.camera.CaptureActivity t45}}} to stack=1 task=45 at 1
I/python  (13908): [INFO              ] [Android     ] Must go into sleep mode, check the app
I/python  (13908): [INFO              ] [Android     ] App doesn't support pause mode, stop.
I/python  (13908): [INFO              ] [Base        ] Leaving application in progress...
I/python  (13908): [INFO              ] [XXXXXXXX    ] --- App STOPPED =======================
```

I've been struggling for a while to get the camera on Android working with Kivy (not easy), and this bug is the latest blocker (otherwise I love kivy!)

Thanks.

**Comment by marystern** [comment]
..by the way, I'm using the code from plyer/platforms/android/camera.py and calling _take_picture() to switch to the camera app and hence enter pause mode ( a simple app with a button that starts the camera using this should suffice to reproduce the bug I think).

**Comment by marystern** [comment]
..I also suspect that it might be the same problem as #3202 , ie that SDL_APP_WILLENTERBACKGROUND is not being handled properly (as per https://wiki.libsdl.org/SDL_EventType )

**Comment by marystern** [comment]
Okay, here's the problem (I think).

The *BASE* class on_pause() method is being called rather than my app's version of that method.

I added debugging code and it seems that the problem is with the shared, class-wide variable "App._running_app". This is set in the App's init method and is *BEING CALLED TWICE**, the first time (correctly) setting that to my subclass, and then later (incorrectly) overriding it to another (base class) object, and hence the on_pause method no longer works as it will get the wrong "app = App.get_running_app()"when it is called from support.py:android_check_pause() and think that pause mode is not supported!

I'm not sure where the second call is coming from, but you can see it in these logs (look for "SETTING" below...:

```
I/python  (24516): [INFO              ] [Kivy        ] v1.9.1-dev
I/python  (24516): [INFO              ] [Python      ] v2.7.2 (default, Aug 20 2015, 00:22:04) 
I/python  (24516): [GCC 4.8]
I/python  (24516): [INFO              ] [Factory     ] 177 symbols loaded
I/python  (24516): [INFO              ] [Image       ] Providers: img_tex, img_dds, img_gif, img_pygame (img_pil, img_ffpyplayer ignored)
I/python  (24516): [INFO              ] [Text        ] Provider: pygame
I/python  (24516): [INFO              ] [Audio       ] Providers: audio_pygame (audio_pygst, audio_ffpyplayer ignored)

I/python  (24516): [INFO              ] [XXXX        ] app.py::App.__init__() ***SETTING App._running_app to self=<xxxxxxxx_app.MyApp object at 0x61694ed8>

I/python  (24516): [INFO              ] [Window      ] Provider: pygame
I/libSDL  (24516): SDL_SetVideoMode(): application requested mode 800x1205
I/python  (24516): [INFO              ] [GL          ] OpenGL version <OpenGL ES 2.0 14.01003>
I/python  (24516): [INFO              ] [GL          ] OpenGL vendor <NVIDIA Corporation>
I/python  (24516): [INFO              ] [GL          ] OpenGL renderer <NVIDIA Tegra 3>
I/python  (24516): [INFO              ] [GL          ] OpenGL parsed version: 2, 0
I/python  (24516): [INFO              ] [GL          ] Texture max size <2048>
I/python  (24516): [INFO              ] [GL          ] Texture max units <16>
I/python  (24516): [INFO              ] [Support     ] Android install hooks
I/python  (24516): [INFO              ] [Window      ] virtual keyboard not allowed, single mode, not docked

I/python  (24516): [INFO              ] [XXXX        ] app.py::App.__init__() ***SETTING App._running_app to self=<kivy.app.App object at 0x617c0a40>

I/python  (24516): [INFO              ] [GL          ] NPOT texture support is available
I/python  (24516): [INFO              ] [xxxxxxxx    ] === App STARTED =======================
I/python  (24516): [INFO              ] [XXXX        ] App.get_running_app()=<kivy.app.App object at 0x617c0a40>
I/python  (24516): [INFO              ] [XXXX        ] self=<xxxxxxxx_app.MyApp object at 0x61694ed8>
```
The App constructor that's called twice
```
    def __init__(self, **kwargs):
        App._running_app = self
        Logger.info('XXX: app.py::App.__init__() ***SETTING App._running_app to self='+str(self))
 ```

**Comment by marystern** [comment]
here's a **workaround** which enables my on_pause method to get called correctly (there are clearly 2 different App objects when running on Android, and the second one breaks the first one. The "proper" fix is to find out where/how the second App object is getting created and then to see what to do with this _running_app variable which assumes there's only ever one App object.):

kivy/app.py:class App(EventDispatcher):
lines 465 onwards:
```
    def __init__(self, **kwargs):
        if App._running_app is None:
            App._running_app = self
            Logger.info('XXXX App: app.py::App.__init__() *** SETTING App._running_app to self='+str(self))
        else:
            # protect against multiple App objects to make on_pause() work!
            Logger.warning('XXXX App: app.py::App.__init__() *** _NOT_ SETTING App._running_app to self='+str(self)+" (because it's already set to App._running_app="+str(App._running_app)+")")

```

Here's a succesful on_pause with my hack, note the WARNING line which shows the hack in action (preventing the App._running_app getting replaced):
```
I/python  ( 3187): [INFO              ] [Image       ] Providers: img_tex, img_dds, img_gif, img_pygame (img_pil, img_ffpyplayer ignored)
I/python  ( 3187): [INFO              ] [Text        ] Provider: pygame
I/python  ( 3187): [INFO              ] [Audio       ] Providers: audio_pygame (audio_pygst, audio_ffpyplayer ignored)

I/python  ( 3187): [INFO              ] [XXXX App    ] app.py::App.__init__() *** SETTING App._running_app to self=<xxxxxxxx_app.XxxXxxxxApp object at 0x6ac92ed8>

I/python  ( 3187): [INFO              ] [Window      ] Provider: pygame
I/libSDL  ( 3187): SDL_SetVideoMode(): application requested mode 800x1205
I/python  ( 3187): [INFO              ] [GL          ] OpenGL version <OpenGL ES 2.0 14.01003>
I/python  ( 3187): [INFO              ] [GL          ] OpenGL vendor <NVIDIA Corporation>
I/python  ( 3187): [INFO              ] [GL          ] OpenGL renderer <NVIDIA Tegra 3>
I/python  ( 3187): [INFO              ] [GL          ] OpenGL parsed version: 2, 0
I/python  ( 3187): [INFO              ] [GL          ] Texture max size <2048>
I/python  ( 3187): [INFO              ] [GL          ] Texture max units <16>
I/python  ( 3187): [INFO              ] [Support     ] Android install hooks
I/python  ( 3187): [INFO              ] [Window      ] virtual keyboard not allowed, single mode, not docked
W/GAV2    ( 3154): Thread[main,5,main]: dispatch call queued.  Need to call GAServiceManager.getInstance().initialize().
I/ActivityManager(  464): Killing 2539:com.google.android.setupwizard/u0a45 (adj 15): empty #17

I/python  ( 3187): [WARNING           ] [XXXX App    ] app.py::App.__init__() *** _NOT_ SETTING App._running_app to self=<kivy.app.App object at 0x6add3ae8> (because it's already set to App._running_app=<xxxxxxxx_app.XxxXxxxxApp object at 0x6ac92ed8>)

I/python  ( 3187): [INFO              ] [GL          ] NPOT texture support is available
I/python  ( 3187): [INFO              ] [xxxxxxxx    ] === App STARTED =======================
I/python  ( 3187): [INFO              ] [XXXX        ] App.get_running_app()=<xxxxxxxx_app.XxxXxxxxApp object at 0x6ac92ed8>
I/python  ( 3187): [INFO              ] [XXXX        ] self=<xxxxxxxx_app.XxxXxxxxApp object at 0x6ac92ed8>
I/python  ( 3187): [INFO              ] [xxxxxxxx    ] <-- App PAUSING .......................
I/python  ( 3187): [INFO              ] [xxxxxxxx    ] Game.game_pause()
I/python  ( 3187): [INFO              ] [OSC         ] using <thread> for socket
I/python  ( 3187): [INFO              ] [Base        ] Start application main loop
I/python  ( 3187): [INFO              ] [Android     ] found 17 joystick
I/python  ( 3187): [INFO              ] [Android     ] create joystick <0> ...
I/ActivityManager(  464): START u0 {act=android.media.action.IMAGE_CAPTURE flg=0x3 cmp=com.google.android.GoogleCamera/com.android.camera.CaptureActivity (has clip) (has extras)} from uid 10156 on display 0
V/WindowManager(  464): addAppToken: AppWindowToken{307e9d80 token=Token{2a804a03 ActivityRecord{b309ab2 u0 com.google.android.GoogleCamera/com.android.camera.CaptureActivity t21}}} to stack=1 task=21 at 1
I/python  ( 3187): [INFO              ] [Android     ] Must go into sleep mode, check the app
I/python  ( 3187): [INFO              ] [XXXX        ] App.get_running_app()=<xxxxxxxx_app.XxxXxxxxApp object at 0x6ac92ed8>
I/python  ( 3187): [INFO              ] [XXXX_XXX    ] kivy.support.py:android_check_pause() calling App.dispatch(on_pause) ******
I/python  ( 3187): [INFO              ] [xxxxxxxx    ] <-- App PAUSING .......................
I/python  ( 3187): [INFO              ] [xxxxxxxx    ] Game.game_pause()
I/python  ( 3187): [INFO              ] [XXXX_XXX    ] kivy.support.py: App.dispatch(on_pause) =True******
I/python  ( 3187): [INFO              ] [xxxxxxxx    ] <-- App PAUSING .......................
I/python  ( 3187): [INFO              ] [xxxxxxxx    ] Game.game_pause()
I/python  ( 3187): [INFO              ] [Android     ] App paused, now wait for resume.
```

**Comment by marystern** [comment]
Now on_pause() is called,but on_resume() is not yet being =called (I get an ANR: Applicton Not Responding due to a timeout even though visually it switches back to my Kivy App
```
E/ActivityManager(  464): ANR in x.x.xxxxxxxx:python (x.x.xxxxxxxxx/org.renpy.android.PythonActivity)
E/ActivityManager(  464): PID: 3187
E/ActivityManager(  464): Reason: Input dispatching timed out (Waiting to send non-key event because the touched window has not finished processing certain input events that were delivered to it over 500.0ms ago.  Wait queue length: 3.  Wait queue head age: 6032.6ms.
```

I guess this is a separate issue.

**Comment by marystern** [comment]
Mystery solved! I have aline like this inside one of my widgets:
```
        snapshot_fname = os.path.join(App().user_data_dir, 'snapshot.png')
```
so, it's my fault because I need the user_data_dir and am creating a second App object just for this. (I remember thinking at the time that I didn't want to create a new object, but there didn't seem to be any other way to get hold of user_data_dir .. I'll investigate: why can't this just be class-wide static data?).

**Comment by marystern** [comment]
Regarding ***on_resume() not being called***, I can see this in traces.txt; does it mean that some part of kivy's SDL code is hanging?

```
"Thread-544" prio=5 tid=16 Waiting
  | group="main" sCount=1 dsCount=0 obj=0x42615460 self=0x5b3ef640
  | sysTid=29828 nice=0 cgrp=default sched=0/0 handle=0x5b3efa18
  | state=S schedstat=( 408874053000 3547913000 125506 ) utm=40073 stm=814 core=1 HZ=100
  | stack=0x6a226000-0x6a228000 stackSize=1036KB
  | held mutexes=
  at java.lang.Object.wait!(Native method)
  - waiting on <0x3ff8479d> (a org.renpy.android.SDLSurfaceView)
  at org.renpy.android.SDLSurfaceView.waitForResume(SDLSurfaceView.java:454)
  - locked <0x3ff8479d> (a org.renpy.android.SDLSurfaceView)
  at org.renpy.android.SDLSurfaceView.nativeInit(Native method)
  at org.renpy.android.SDLSurfaceView.run(SDLSurfaceView.java:730)
  at java.lang.Thread.run(Thread.java:818)

```

**Comment by marystern** [comment]
FYI: re-tested with KivyLauncher (as suggested by qua-non on IRC), but same result: on_resume() is not caled at all :(

**Comment by marystern** [comment]
FYI: A simple app works (on_resume() gets called), including with the camera, so there's something in the full app causing the problem it seems (it's a game with images, sounds, a few Screens and a timer: pretty simple). TODO: I'll drill down more and see if I can find the cause.

**Comment by Ramalus** [comment]
I'm developing a app and testint it on Android 5.1 (cyanogenmod). I'll try
to implement the on_resume method and see if I can input any usefull
information here.

Em seg, 24 de ago de 2015 às 19:33, marystern <notifications@github.com>
escreveu:

**Comment by marystern** [comment]
Thanks Ramalus.

FYI: I added some debuging in PythonActivity.java and I can see the onPause() method being called, but the activity's onResume() method does NOT get called.

**Comment by marystern** [comment]
I've added some logging and noticed the following. ***AFTER onPause has finished, the SDLSurfaceView: receives a touch event***. Presumably this event gets queued, but never handled, and hence the ANR and lack of onResume() callback by Android.
```
I/python  (15189): [INFO              ] [xxxxxxxx    ] <-- App PAUSING .......................
I/python  (15189): [INFO              ] [Android     ] App paused, now wait for resume.
I/xxxx_p4a(15189): SDLSurfaceView: >>> waitForResume() ENTER ------------------
I/xxxx_p4a(15189): SDLSurfaceView: --- waitForResume() ... about to notifyAll() threads waiting onPause
I/xxxx_p4a(15189): SDLSurfaceView: --- waitForResume() ... about to notifyAll() DONE, about to loop waiting for resume...
I/xxxx_p4a(15189): SDLSurfaceView: --- waitForResume() ... looping: BEFORE wait()...
I/xxxx_p4a(15189): SDLSurfaceView: <<< onPause() EXIT ------------------
I/xxxx_p4a(15189): <<< onPause() EXIT ------------------
I/ActivityManager(  466): Start proc 16141:com.google.android.GoogleCamera/u0a138 for activity com.google.android.GoogleCamera/com.android.camera.CaptureActivity
I/xxxx_p4a(15189): SDLSurfaceView: ... onTouchEvent(event=MotionEvent { action=ACTION_UP, id[0]=0, x[0]=214.06575, y[0]=795.46405, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=9680679, downTime=9680598, deviceId=2, source=0x1002 }) ------------------
```

**Comment by akshayaurora** [comment]
""FYI: A simple app works (on_resume() gets called), including with the
camera, so there's something in the full app causing the problem it seems""


Reduce your game/app step by app see what step causes the issue,  it's
clearly a issue with something you do in your app, maybe you don't pause
the sound/video in on_pause, maybe you do a operation that blocks the app
from pausing properly

**Comment by marystern** [comment]
@akshayaurora yes, I was trying that, but haven't found anything yet.

Can you tell me what things I'm supposed to do myself vs kivy handling things please (I'd assumed that kivy would handle most things for me: at the point of pause I'm unaware of anything major that is running in the background: there is no audio or video running)?

Also, can anyone plese tell me ***if the app shuld still receive events after being paused*** please (ie is it designed to handle that, or is that a problem)? -( see previous comment)

**Comment by marystern** [comment]
For info, here's the SDLSurfaceView thread waiting for resume (after it sould have been resumed, ie while it is not responding) - from android's traces.txt log:
```
"Thread-384" prio=5 tid=14 Waiting
  | group="main" sCount=1 dsCount=0 obj=0x12cfb340 self=0x60ff5800
  | sysTid=6518 nice=0 cgrp=default sched=0/0 handle=0x60ff5ce0
  | state=S schedstat=( 5110592000 118676000 859 ) utm=487 stm=24 core=3 HZ=100
  | stack=0x69d03000-0x69d05000 stackSize=1036KB
  | held mutexes=
  at java.lang.Object.wait!(Native method)
  - waiting on <0x3a41e61c> (a org.renpy.android.SDLSurfaceView)
  at org.renpy.android.SDLSurfaceView.waitForResume(SDLSurfaceView.java:476)
  - locked <0x3a41e61c> (a org.renpy.android.SDLSurfaceView)
  at org.renpy.android.SDLSurfaceView.nativeInit(Native method)
  at org.renpy.android.SDLSurfaceView.run(SDLSurfaceView.java:759)
  at java.lang.Thread.run(Thread.java:818
```

**Comment by Ramalus** [comment]
I've just tested this im my app and it works as it should.
This is the relevant log.

I/python  (23361): [INFO              ] [Android     ] Must go into sleep mode, check the app
I/python  (23361): do some stuff on_pause
I/python  (23361): [INFO              ] [Android     ] App paused, now wait for resume.
I/ActivityManager(  783): START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=org.ramalus.caruni/org.renpy.android.PythonActivity bnds=[536,292][712,522] (has extras)} from uid 10019 on display 0
I/python  (23361): [INFO              ] [Android     ] Android has resumed, resume the app.
I/python  (23361): do something on resume
I/python  (23361): [INFO              ] [Android     ] App resume completed.
I/Python: (23361): Layout Triggered, Keyboard_height: 50
I/Timeline(  783): Timeline: Activity_windows_visible id: ActivityRecord{127a0d95 u0 org.ramalus.caruni/org.renpy.android.PythonActivity t3459} time:72780816

**Comment by marystern** [comment]
@Ramalus - thanks: I'll dig deeper into my app and try to isolate the problem. By the way, this process is very slow as buildozer seems to require 2 or 3 goes to build properly (regulalry fails first time frm clean with ant debug failure , and even when its stabilised, it wants to re-build python every time I make a code change..annoyingly slow and unnecessary..shame :(

Error: Target id 'android-14' is not valid followed by ant debug not found error.

..usually cleaning/distcleaning and re-trying eventually sorts it out :(

**Comment by Ramalus** [comment]
Hum, how slow are you talking about? When the change isn't massive and doesn't include several other libraries its not that slow after the first time at least for me.
Well it does take about 30 seconds but that's acceptable I guess.

**Comment by marystern** [comment]
***Mystery SOLVED!!***

In my code I was using Android's accelerometer (plyer.accelerometer) and I called enable() on it, but apparently this must be disable()'d during pause!

..I don't fully understand why..but it makes on_resume() work again at least!
